### ISSUE TITLE
rr.com mail issue

### ISSUE THREAD

**Issue by tree1891** [jstedfast/MailKit#762]
I can see a new mail in https://webmail.roadrunner.com, but it is not in mailkit imap inbox.
If mail client reconnect, it is in inbox.

**Comment by jstedfast** [comment]
Is this via POP3? If so, this is the way that POP3 servers work. They "freeze" their "view" if the Inbox when you connect and do not update it again until the POP3 client reconnects. The reason for this is that there is no way for the server to signal to the client that anything has changed and the protocol is too basic (i.e. it uses indexes rather than unique identifiers) to be usable if the content was changing out from under the client.

**Comment by tree1891** [comment]
It is via IMAP.

**Comment by jstedfast** [comment]
Does the server notify MailKit that a new message exists?

**Comment by tree1891** [comment]
I don't use any callback.

**Comment by jstedfast** [comment]
That doesn't matter.

**Comment by tree1891** [comment]
OpenPop seems to work.

**Comment by jstedfast** [comment]
OpenPop isn't IMAP.

**Comment by jstedfast** [comment]
Please follow these instructions: https://github.com/jstedfast/MailKit/blob/master/FAQ.md#ProtocolLog

**Comment by tree1891** [comment]
I know, but Mailkit pop3 has same issue.

**Comment by jstedfast** [comment]
I don't believe you. Show me protocol logs.

**Comment by tree1891** [comment]
First tried 5 times and after reconnect, also 5 times, but not get new mail.
In webmail page, I can see new mail.
[imap.zip](https://github.com/jstedfast/MailKit/files/2429044/imap.zip)

**Comment by tree1891** [comment]
In other mail servers, it does work, but not only in rr.com.

**Comment by jstedfast** [comment]
I want you to take a look at the imap.log file in a text editor.

It looks to me like you are explicitly trying to open the INBOX folder each time before downloading a message. This should be unnecessary and indeed, each time the folder is opened, it claims to have exactly the same number of messages in the INBOX:

```
C: A00000005 SELECT INBOX
S: * 2032 EXISTS
S: * OK [UNSEEN 57] First unseen message
S: * OK [UIDVALIDITY 1238157015] UIDs valid
S: * OK [UIDNEXT 23655] Predicted next UID
S: * FLAGS ($Forwarded \Answered \Flagged \Deleted \Draft \Seen)
S: * OK [PERMANENTFLAGS ($Forwarded \* \Answered \Flagged \Deleted \Draft \Seen)] Permanent flags
S: * 2 RECENT
S: A00000005 OK [READ-WRITE] SELECT completed
...
C: A00000007 SELECT INBOX
S: * 2032 EXISTS
S: * OK [UNSEEN 57] First unseen message
S: * OK [UIDVALIDITY 1238157015] UIDs valid
S: * OK [UIDNEXT 23655] Predicted next UID
S: * FLAGS ($Forwarded \Answered \Flagged \Deleted \Draft \Seen)
S: * OK [PERMANENTFLAGS ($Forwarded \* \Answered \Flagged \Deleted \Draft \Seen)] Permanent flags
S: * 0 RECENT
S: A00000007 OK [READ-WRITE] SELECT completed
...
C: A00000009 SELECT INBOX
S: * 2032 EXISTS
S: * OK [UNSEEN 57] First unseen message
S: * OK [UIDVALIDITY 1238157015] UIDs valid
S: * OK [UIDNEXT 23655] Predicted next UID
S: * FLAGS ($Forwarded \Answered \Flagged \Deleted \Draft \Seen)
S: * OK [PERMANENTFLAGS ($Forwarded \* \Answered \Flagged \Deleted \Draft \Seen)] Permanent flags
S: * 0 RECENT
S: A00000009 OK [READ-WRITE] SELECT completed
...
C: A00000011 SELECT INBOX
S: * 2032 EXISTS
S: * OK [UNSEEN 57] First unseen message
S: * OK [UIDVALIDITY 1238157015] UIDs valid
S: * OK [UIDNEXT 23655] Predicted next UID
S: * FLAGS ($Forwarded \Answered \Flagged \Deleted \Draft \Seen)
S: * OK [PERMANENTFLAGS ($Forwarded \* \Answered \Flagged \Deleted \Draft \Seen)] Permanent flags
S: * 0 RECENT
S: A00000011 OK [READ-WRITE] SELECT completed
...
C: A00000013 SELECT INBOX
S: * 2032 EXISTS
S: * OK [UNSEEN 57] First unseen message
S: * OK [UIDVALIDITY 1238157015] UIDs valid
S: * OK [UIDNEXT 23655] Predicted next UID
S: * FLAGS ($Forwarded \Answered \Flagged \Deleted \Draft \Seen)
S: * OK [PERMANENTFLAGS ($Forwarded \* \Answered \Flagged \Deleted \Draft \Seen)] Permanent flags
S: * 0 RECENT
S: A00000013 OK [READ-WRITE] SELECT completed
```

**Comment by tree1891** [comment]
I have checked inbox every 10 seconds until find new.

**Comment by jstedfast** [comment]
Now for a suggestion that *may* solve your problem.

IMAP has a `CHECK` command that forces the IMAP server to resynchronize its in-memory and on-disk representations of the currently opened folder.

Try calling `client.Inbox.Check()` or `await client.Inbox.CheckAsync()` to see if that updates the Count.

I would also recommend not opening the folder multiple times as that is an expensive server operation and is unnecessary.

**Comment by jstedfast** [comment]
> I have checked inbox every 10 seconds until find new.

Right, but clearly the server isn't telling MailKit that new messages exist. That's the problem. The problem does not appear to be a bug in MailKit that I can find based on the protocol logs.

**Comment by tree1891** [comment]
I have used Check(), but does not help resynch, so I use reopening inbox.

**Comment by jstedfast** [comment]
Does re-opening the INBOX work?

**Comment by tree1891** [comment]
it does work on other mail servers, not rr.com.

**Comment by jstedfast** [comment]
I'm not sure what to suggest, but there's no bug in MailKit's IMAP client implementation (obviously MailKit cannot know that new messages exist unless the server tells us about them and the server isn't doing that).

What happens if you select a different folder and then switch back to the Inbox? Does that cause the server to maybe recognize that the Inbox contains new messages?

**Comment by tree1891** [comment]
But in utility using OpenPop, I can see new mail.

**Comment by jstedfast** [comment]
That's irrelevant. POP3 and IMAP function completely differently.

**Comment by tree1891** [comment]
I know, but MailKit's POP3 also has same issue in rr.com

**Comment by jstedfast** [comment]
There is a legitimate design problem with MailKit's POP3 implementation in that MailKit queries for the message count after authenticating and doesn't provide a way to update that message count. I can add an API to do this, but that doesn't have any effect on IMAP.

**Comment by tree1891** [comment]
If pop3 does work, I can use it on rr.com

**Comment by tree1891** [comment]
# What happens if you select a different folder and then switch back to the Inbox?
How to do it?

**Comment by jstedfast** [comment]
I don't know what folders you have other than your Inbox, but it's just like opening the Inbox folder, except you open a different folder and then re-open the Inbox folder again.

Most IMAP servers have at least a Drafts folder and a Sent folder (altho this one can be named "Sent", "Sent Messages", "sent-messages", "sent-mail", etc depending on the IMAP server so it'd be hard to guess).

```csharp
var personal = client.GetFolder (client.PersonalNamespaces[0]);
var drafts = personal.GetSubfolder ("Drafts");
drafts.Open (FolderAccess.ReadWrite);
client.Inbox.Open (FolderAccess.ReadWrite);
...
```

**Comment by jstedfast** [comment]
I should note that different IMAP servers may put the Drafts folder at the root and some might put it under Inbox, so you'll have to figure out which behavior your rr.com IMAP server does.

**Comment by tree1891** [comment]
It does not help.

**Comment by tree1891** [comment]
Is there any pop3 cache?
After reconnect two times, pop client can read new mail, but not OpenPOP.
Is there any difference between reconnect and creating instance?

**Comment by jstedfast** [comment]
I'm confused by your comment.

> After reconnect more than 2 times, pop client can read new mail, but not OpenPOP.

Did you mean that MailKit's Pop3Client can read new mail, but OpenPOP's Pop3Client cannot?

Or did you typo and mean the reverse?

> Is there any difference between reconnect and creating instance?

For MailKit's Pop3Client, there should be no difference between creating a new Pop3Client vs re-connecting an existing Pop3Client.

I'm not sure about OpenPOP.

**Comment by tree1891** [comment]
If there exist new mail, OpenPOP can read it without reconnecting, but not MailKit.

**Comment by jstedfast** [comment]
If you build MailKit from source, that issue is fixed (you'll need to use the `Pop3Client.GetMessageCount[Async]()` method instead of using the `Count` property, though.

See here for the fix: https://github.com/jstedfast/MailKit/commit/00cae6aaccefeca2c0d170a04519d17f0dd31edc

**Comment by tree1891** [comment]
But I don't think so, because new mail can't be read after reconnect only one time.
Sometimes it needs more than 3 times reconnecting.

**Comment by tree1891** [comment]
If count update is problem, new mail must be read after reconnect once.

**Comment by jstedfast** [comment]
You're not making any sense at all.

**Comment by jstedfast** [comment]
You do not need to reconnect in order to call `GetMessageCount()`. You can call `GetMessageCount()` any time you want.

```csharp
using (var client = new Pop3Client ()) {
    client.Connect ("pop3.server.com", 110, false);
    client.Authenticate ("username", "password");

    int count = client.GetMessageCount ();
    for (int i = 0; i < count; i++) {
        var message = client.GetMessage (i);
    }

    // wait a few minutes, check if there are any new messages
    int newCount = client.GetMessageCount ();
    if (newCount > count) {
        // new messages have arrive, download them
        for (int i = count; i < newCount; i++) {
            var message = client.GetMessage (i);
        }
    }
}
```

**Comment by tree1891** [comment]
I mean the case that there exists new mail before MailKit authenticate.
In that case, OpenPOP authenticate and can read it, but not MailKit.

**Comment by jstedfast** [comment]
You still aren't making any sense.

If you aren't authenticated, MailKit can't even get the message count - so what difference does it make if new mail arrives when you aren't authenticated?

OpenPOP can't check the message count while it isn't authenticated either.

It sounds to me like you have no idea what you are talking about and are just spouting gibberish.

**Comment by tree1891** [comment]
At the beginning, I said that I can see unread mail in webmail, but MailKit can't read it.
Do you understand?

**Comment by jstedfast** [comment]
Yes, I understand that.

But it doesn't matter if new mail arrives in your inbox when MailKit isn't connected and/or authenticated - it doesn't even request the message count until after you are authenticated (it can't - the POP3 server won't let it).

Let me give you a little scenario:

- You have 10 messages in your Inbox on the mail server.
- You create an instance of MailKit's Pop3Client.
- You call `client.Connect (...);`
- A new message arrives in your Inbox.

Here's where you are telling me "but MailKit can't see the new message at this point!"

Well, no shit Sherlock. It hasn't queried the message count yet.

If you are using MailKit 2.0.6, it will query the message count once you call `Authenticate()` and it will store the message count in the `Count` property.

Ok, so let's continue this scenario...

- You call `client.Authenticate (...);`
- You check `client.Count` and the value will be `11`.
- You loop over the messages from `0` to `10` (`for (int i = 0; i < 11; i++)`)
- MailKit successfully downloads those messages.
- A new message arrives in your Inbox
- MailKit 2.0.6's `Pop3Client.Count` property will still say `11` because it only checks this value once during the `Authenticate (...)` method call. Luckily for you, I added `GetMessageCount()` which, if you call it, will return a value of `12`.

**Comment by tree1891** [comment]
I'll continue.
...
- A new message arrives in your Inbox
- Pop3Client.Count property will still say 11
- disconnect and client.Authenticate (...);
- Pop3Client.Count property will still say 11, not 12.

That is problem.

**Comment by jstedfast** [comment]
Show me a protocol log because that is impossible unless the server tells MailKit that there are only 11 messages instead of 12.

Look at the code for yourself if you don't believe me. When MailKit disconnects, it sets Count to 0. So it can't possibly cache that value between sessions.

**Comment by tree1891** [comment]
Please check this.
[pop3.log](https://github.com/jstedfast/MailKit/files/2469962/pop3.log)

**Comment by tree1891** [comment]
```
...
                client.Connect(host, port, useSsl);
                client.Authenticate(username, password);

                Console.WriteLine("Hit Return when new mail arrives");
                Console.ReadLine();

                client.Disconnect(true);

                client.Connect(host, port, useSsl);
                client.Authenticate(username, password);
                client.Disconnect(true);
...
```
[pop3.log](https://github.com/jstedfast/MailKit/files/2470005/pop3.log)

**Comment by jstedfast** [comment]
```
...
C: STAT
S: +OK 46 16676080
...
C: STAT
S: +OK 46 16676080
...
```

Your mail server is reporting a message count of `46` in both sessions. This is not a bug in MailKit.

**Comment by tree1891** [comment]
Then How does OpenPOP work?

**Comment by jstedfast** [comment]
Probably you are just getting lucky with your timing.

There's a fundamental thing you need to understand with the way mail works. Just because you can see a message in your Inbox using the webmail client does not mean that it is available using IMAP or POP3. Sometimes there is a delay.

Did you notice that in your first log (from this comment: https://github.com/jstedfast/MailKit/issues/762#issuecomment-429043280), you *did* get an increase in the message count after a while? (it was originally 42, then you deleted a message, then you reconnected and it was 41, then it stayed 41 for a few reconnects, and then it became 42 again when a new message arrived).

The reason it finally went to 42 is not because you reconnected the magic number of times. The reason it went to 42 is because that message was finally delivered to the POP3 server after a delayed period of time.

**Comment by tree1891** [comment]
Maybe you are right, but OpenPOP is also pop client.

**Comment by jstedfast** [comment]
Yes, and I guarantee it does exactly the same thing. In fact, I can prove that it does: https://github.com/foens/hpop/blob/master/OpenPop/Pop3/Pop3Client.cs#L438

**Comment by jstedfast** [comment]
> It detects new message before Mailkit try to reconnect.

*sigh*

Try to keep up.

We've already established the FACT that your mail server is failing to report the correct number of messages and that MailKit and OpenPOP use the exact same POP3 command to get the number of messages on the server.

This means that both MailKit and OpenPOP are doing the exact same thing.

This means that if either client sends a "STAT" command, the server will return 46 even when in your webmail client, it says 47.

This means that both MailKit and OpenPOP will behave identically.

The fact that you think otherwise demonstrates that you do not understand how things work. This is especially frustrating to me since I have explained what is happening multiple times and you refuse to listen.

Now...

OpenPOP *does* have an advantage over MailKit because you would not need to reconnect using OpenPOP since you can just call `GetMessageCount()` and each time you call `GetMessageCount()`, OpenPOP will re-send the "STAT" command.

MailKit 2.0.6, however, does not have a `GetMessageCount()` method and so you must disconnect and then reconnect.

That said, I added a `GetMessageCount()` method where each time you call it, it sends a "STAT" command just like OpenPOP so that you will not have to disconnect and reconnect again.

Do you understand?

Or do I have to explain it yet again?

**Comment by tree1891** [comment]
Sorry, OpenPOP client I have used use different server from MailKit client.

**Comment by jstedfast** [comment]
😄

**Comment by jstedfast** [comment]
I have released MailKit 2.0.7 that has the GetMessageCount() and GetMessageCountAsync() methods.

**Comment by tree1891** [comment]
Thanks
